### PR TITLE
Implement Goal Driver

### LINKED ISSUE #85: Driver: Goal Detection

### PR DESCRIPTION
**Pull request by tream2001**
~~The driver's design centers around a ScoreState enum which stores whether either player has scored or not.~~

~~When a player scores, an external interrupt sets ScoreState to the appropriate state. ScoreState is later checked by the code handling the game play to switch between states appropriately.~~

The driver implements a function for polling each of the goal detection circuit outputs.

closes #85

### CODE REVIEW

**Comment by WDobert** [comment]
My bad for not mentioning sooner, might be good to shorten the name to just `driver_goal_...` to keep it succinct

File: firmware-nucleo/smart-air-hockey-table/App/Inc/driver_goaldetection.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#define INC_DRIVER_GOALDETECTION_H_
+
+#include <stdbool.h>
+#include <inttypes.h>
+
+typedef enum _ScoreState {
+	ScoreState_Clear,	// Neither player has scored
+	ScoreState_A,		// Player A has scored
+	ScoreState_B		// Player B has scored
+} ScoreState;
```
Comment (alanssitis) [inline firmware-nucleo/smart-air-hockey-table/App/Inc/driver_goaldetection.h:11]: Wouldn't we rather make the change directly to game state? i.e. calling
```
App_StateMachine_SetState(GameState_PlayerAScored)
```
[Look here](https://github.com/alanssitis/smart-air-hockey-table/pull/87#issuecomment-1765547660)

This way we can have all logic be handled by a single entity.

**Review by alanssitis — COMMENTED** [review]
I feel like trying to keep track of multiple states will be a bit annoying and could cause issues down the line. Might make the most sense for a single entity to track it.

File: firmware-nucleo/smart-air-hockey-table/App/Inc/driver_goaldetection.h (hunk 1 above)
Comment (tream2001) [reply firmware-nucleo/smart-air-hockey-table/App/Inc/driver_goaldetection.h:11]: I was worried about issues with multiple interrupts being raised, especially if we are trying to distinguish between the score and win states.

Comment (alanssitis) [reply firmware-nucleo/smart-air-hockey-table/App/Inc/driver_goaldetection.h:11]: you can go from scored -> change score -> change to appropriate state.
else, I think we can be very prone to race conditions.

**Review by WDobert — CHANGES_REQUESTED** [review]
Suggested change for consistency with other local variables names in the project.
Recommend using `LL_` functions to increase readability, but I understand if the preference is to stick to CMSIS.

**Comment by tream2001** [comment]
> Recommend using `LL_` functions to increase readability, but I understand if the preference is to stick to CMSIS.

The most relevant `LL_` function would be "LL_GPIO_IsInputPinSet", but I am not sure of the performance implications. It would also require the use of a "GPIO_TypeDef *" variable, which seems overkill. If you know of a better `LL_` function that can replicate masking the register, then I would be interested to hear it.

**Review by WDobert — APPROVED** [review]

**Review by WDobert — APPROVED** [review]

**Review by alanssitis — APPROVED** [review]

**Review by benjamin-owen — APPROVED** [review]

**Pull request merged**